### PR TITLE
Validate runtime image environments in CI testing

### PR DESCRIPTION
**Pull request by kevin-bates**
This pull request adds an additional GitHub action named "Validate Image Environment" which runs for each supported version of Python and validates the [`requirements-elyra.txt`](https://github.com/elyra-ai/elyra/blob/master/etc/generic/requirements-elyra.txt) and requirements-elyra-py37.txt (in this PR) files that consists of the package dependencies used in the generic component containers can be successfully installed relative to the corresponding version of Python.
This also modifies the validate-runtime-image target to validate the elyra sample runtime images against these requirements

Resolves: #2719
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -318,8 +318,14 @@ validate-runtime-images: # Validates delivered runtime-images meet minimum crite
...
 			fi; \
+			if [ $$cmd == "python3" ]; then \
+				IMAGE_PYTHON3_MINOR_VERSION=`docker run --rm $$image $$cmd --version | cut -d' ' -f2 | cut -d'.' -f2` ; \
+				if [[ $$IMAGE_PYTHON3_MINOR_VERSION -lt 8 ]]; then \
+					echo WARNING: Image $$image requires at Python 3.8 or greater for generic component dependency installation; \
+				fi; \
+			fi; \
 		done; \
-		docker run -v $(pwd)/etc/generic:/opt/elyra/ --rm $$image python3 -m pip install -r /opt/elyra/requirements-elyra.txt > /dev/null 2>&1 ; \
+		docker run -v $$GITHUB_WORKSPACE/etc/generic:/opt/elyra/ --rm $$image python3 -m pip install -r /opt/elyra/requirements-elyra.txt > /dev/null ; \
```
Comment (kevin-bates) [inline Makefile:328]: What is the default value for `GITHUB_WORKSPACE`?  Seems like we should have something like:
```
GITHUB_WORKSPACE?=$(pwd)
```
so this target can still be used outside of GH actions.

Comment (akchinSTC) [reply Makefile]: Updated, didnt want to include a one off GITHUB* in the makefile..kinda weird, so I just replaced the env directly

**Comment by akchinSTC** [comment]
Ive added second version of the requirements file for ipython <8 (e.g. py37 support)for generic components, which is essentially the same as the original with after reverting the updates to it. This would mean vulnerabilities that we attempted to close in #2491 -> #2487 would exist for this version of the file.

The original requirements file has been updated with bumps for nbclient for the original issue, and a bump for pyzmq due to failures in testing with python 3.10 env and the conda runtime image due to missing or incompatible native compilation libraries (gcc).

Follow up PR will include operator changes pull in both requirements files, bootstrapper would change to detect which version is running and pull in the appropriate version of the requirements file.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -171,6 +171,18 @@ jobs:
       - name: Build
         run: make docs
 
+  validate-image-env:
+    name: Validate Image Environment
```
[the remark this answers is not in the document]
Comment (akchinSTC) [reply .github/workflows/build.yml:175]: > General comment: Could we start adding descriptive comments to 'targets' to make it clearer what their purpose is? For example, as is looking at this file it's not very clear what the difference is between `validate-image-env` and `validate-image`. With descriptions we can avoid that:
> 
> ```
> validate-image-env:
>     # Validate that Elyra's requirements for generic component processing can be installed in container images, which are included as system-managed runtime image configurations.
>     ...
> validate-images:
>     # Validate that included system-managed runtime images meet Elyra's requirements documented in https://elyra.readthedocs.io/en/latest/user_guide/runtime-image-conf.html#prerequisites
> ```

The targets have been renamed respectively with their descriptions
```
elyra-image-env: ## Creates a conda env consisting of the dependencies used in images
validate-runtime-images: # Validates delivered runtime-images meet minimum criteria

```

**Comment by akchinSTC** [comment]
> Is what is listed address the vulnerabilities comment?

Yes looks like those versions would resolve the vulnerabilities. Updated the requirements file to match. 

Pro-active...should we also do the same for the 3.8+ version and bump the associated requirements?

**Comment by kevin-bates** [comment]
> Pro-active...should we also do the same for the 3.8+ version and bump the associated requirements?

This seems reasonable provided we can re-test things. It would be nice to avoid introducing per-version requirements for 3.8+, so perhaps using the versions found against 3.8 will be sufficient (and until we learn otherwise). 

Also, in visiting each of the capped packages in [`setup.py`](https://github.com/elyra-ai/elyra/blob/master/setup.py), it looks like we'd be violating the caps for `jsonschema`, `pyyaml`, and `typing-extensions`.  These caps are due to KFP and I'm not sure if those same caps should be in play for the image requirements, but wanted to point that out.  Should these also apply to the image?  Since we're not using the KFP API within the container (at least that I'm aware), I'm not sure the caps are necessary.

**Comment by akchinSTC** [comment]
> Since we're not using the KFP API within the container (at least that I'm aware), I'm not sure the caps are necessary.

Pretty sure they dont apply. We dont use any kfp libraries to my knowledge in the generic component execution

File: Makefile [hunk 3]
```diff
@@ -23,16 +23,23 @@
 .PHONY: docs-dependencies docs
 .PHONY: elyra-image publish-elyra-image kf-notebook-image publish-kf-notebook-image
```
Comment (ptitzler) [inline Makefile:24]:
```suggestion
.PHONY: elyra-image elyra-image-env publish-elyra-image kf-notebook-image publish-kf-notebook-image
```

**Review by ptitzler — APPROVED** [review]
LGTM

**Pull request merged**